aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -0,0 +1,245 @@
+        // Copy into the new format, contiguous, on the source device
+        src_contig = at::empty_like(dst, src.options().dtype(dst.dtype()));
+
+        CopyOp<dst_T, src_T>::apply(dst, src);
+      }
+
+      // Make sure the dst is contiguous
+      DeviceGuard device_guard_dst{dst_device};
+      Tensor dst_contig = dst.contiguous();
+
+      // Now, we are ready for a cross-device memcpy of contiguous
+      // data, of the same layout and type
+      DeviceGuard device_guard_src{src_device};
+
+      AT_CUDA_CHECK(cudaMemcpyAsync(
+          dst_contig.data<dst_T>(),
+          src_contig.data<dst_T>(),
+          numel * sizeof(dst_T),
+          cudaMemcpyDeviceToDevice,
+          copy_stream));
+
+      if (!dst.is_contiguous()) {
+        copy_device_to_device<dst_T, dst_T>(dst_contig, dst);
+      }
+    }
+  }
+
+  if (src_device != dst_device && copy_stream == NULL) {
+    // dst waits on src barrier (dst already waits on dst). We cannot
+    // operate on dst's copy until the copy is complete.
+
+    // Still on src_device, record default stream event
+    CUDAEvent src_ready;
+    src_ready.record(copy_stream);
+
+    DeviceGuard device_guard{dst_device};
+    src_ready.block(getDefaultCUDAStream(dst_device.index()));
+  }
+
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void copy_from_cpu(Tensor& dst, const Tensor& src) {
+  Tensor dst_contig = dst.contiguous();
+  Tensor src_contig = src.contiguous();
+
+  CUDAStream stream = getCurrentCUDAStream();
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(src.type(), "copy_from_cpu", [&]() {
+    AT_CUDA_CHECK(cudaMemcpyAsync(
+        dst_contig.data<scalar_t>(),
+        src_contig.data<scalar_t>(),
+        src.numel() * sizeof(scalar_t),
+        cudaMemcpyHostToDevice,
+        stream));
+    AT_CUDA_CHECK(cudaStreamSynchronize(stream));
+    copy_device_to_device<scalar_t, scalar_t>(dst, dst_contig);
+  });
+}
+
+void copy_to_cpu(Tensor& dst, const Tensor& src) {
+  Tensor dst_contig = dst.contiguous();
+  Tensor src_contig = src.contiguous();
+
+  DeviceGuard device_guard{src.device()};
+  CUDAStream stream = getCurrentCUDAStream();
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(src.type(), "copy_to_cpu", [&]() {
+    AT_CUDA_CHECK(cudaMemcpyAsync(
+        dst_contig.data<scalar_t>(),
+        src_contig.data<scalar_t>(),
+        src.numel() * sizeof(scalar_t),
+        cudaMemcpyDeviceToHost,
+        stream));
+    AT_CUDA_CHECK(cudaStreamSynchronize(stream));
+    _copy_same_type_(dst, dst_contig);
+  });
+}
+
+template <typename dst_T>
+void _copy__cuda(Tensor& dst, const Tensor& src) {
+  AT_DISPATCH_ALL_TYPES_AND_HALF(src.type(), "_copy__cuda", [&]() {
+    if (dst.is_cuda() && src.is_cuda()) {
+      copy_device_to_device<dst_T, scalar_t>(dst, src);
+    } else if (dst.is_cuda()) {
+      if (std::is_same<dst_T, scalar_t>::value) {
+        copy_from_cpu(dst, src);
+      }
+      Tensor srcf = at::empty_like(src, src.options().dtype(dst.dtype()));
+      _copy_(srcf, src);
```

Comment:
Would appreciate a comment here saying this is a same-device, dtype-converting copy. It's otherwise a little difficult to puzzle out why `_copy__cuda` is calling `_copy_`.